### PR TITLE
Feature/tcp socket implement

### PR DESCRIPTION
**Pull request by yjfan**
Please help to review this change for TCP socket implementation.

### CODE REVIEW

File: src/domain.h [hunk 1]
```diff
@@ -7,11 +7,14 @@ extern "C" {
 
 typedef enum
 {
-    GS_SOCKET_DOMAIN_UNIX = 0
+    GS_SOCKET_DOMAIN_UNIX = 0,
+    GS_SOCKET_DOMAIN_TCP,
+    GS_SOCKET_DOMAIN_UDP,
```
Comment (yjfan) [inline src/domain.h]: Please remove GS_SOCKET_DOMAIN_UDP because we don't support that.

File: src/socket.c [hunk 2]
```diff
@@ -12,6 +13,9 @@ struct gs_socket_t * gs_socket_create(GS_SOCKET_DOMAIN_TYPE domain)
         case GS_SOCKET_DOMAIN_UNIX:
             base = gs_unix_socket_base();
             break;
+    case GS_SOCKET_DOMAIN_TCP:
+        base = gs_tcp_socket_base();
+        break;
```
Comment (yjfan) [inline src/socket.c]: Please let them be the same alignment as #12~#15.

File: tests/receiver.c [hunk 3]
```diff
@@ -216,16 +221,35 @@ int main(int argc, char *argv[])
...
 
+    if (strcspn(address, "://") > 3) {
+        print_usage(argv[0]);
+        exit(EXIT_FAILURE);
+    }
+
+    if (strncmp(address, "TCP", 3) == 0) {
+        type = GS_SOCKET_DOMAIN_TCP;
+    }
+    else if (strncmp(address, "UDP", 3) == 0) {
```
Comment (yjfan) [inline tests/receiver.c]: Please remove the code which related to GS_SOCKET_DOMAIN_UDP.

File: tests/receiver.c [hunk 4]
```diff
@@ -25,6 +26,9 @@ static void print_usage(const char *binary_name)
     const char *format = "Usage: %s -b <address>\n"
                          "Options:\n"
                          "  -b    bind and listen the address\n"
+                         "        address format example :\n"
+                         "        TCP://[ip addr]:[port]\n"
+                         "        IPC://[path]\n"
```
Comment (yjfan) [inline tests/receiver.c]: Suggest that you should use lower-case for protocol. (e.g., tcp://<address>:<port>)
Do not use square brackets that mean the arguments might be omitted.

File: tests/sender.c [hunk 5]
```diff
@@ -15,15 +16,18 @@ static void print_usage(const char *binary_name)
     const char *format = "Usage: %s -c <address> -m <message>\n"
                          "Options:\n"
                          "  -c    connect to the remote address\n"
+                         "        address format example :\n"
+                         "        TCP://[ip addr]:[port]\n"
+                         "        IPC://[path]\n"
```
Comment (yjfan) [inline tests/sender.c]: The same as tests/receiver.c #30 ~ #31.

File: tests/sender.c [hunk 6]
```diff
@@ -7,6 +7,7 @@
 #include <errno.h>
 
 #include "../src/gs.h"
+#include "../src/domain.h"
```
Comment (yjfan) [inline tests/sender.c]: Redundant line.

File: tests/receiver.c [hunk 7]
```diff
@@ -11,6 +11,7 @@
 #include <pthread.h>
 
 #include "../src/gs.h"
+#include "../src/domain.h"
```
Comment (yjfan) [inline tests/receiver.c]: Redundant line.

File: tests/sender.c [hunk 8]
```diff
@@ -82,10 +87,29 @@ int main(int argc, char *argv[])
...
+    }
+    else if (strncmp(address, "IPC", 3) == 0) {
+        type = GS_SOCKET_DOMAIN_UNIX;
+    }
+    else {
+        print_usage(argv[0]);
+        exit(EXIT_FAILURE);
+    }
+
+    send_message(type, strstr(address, "://") + 3, message, strlen(message));
```
Comment (yjfan) [inline tests/sender.c]: Consider changing to make is as beautiful as possible.
e.g.,
```c
GS_SOCKET_DOMAIN_TYPE types[] = {GS_SOCKET_DOMAIN_TCP, GS_SOCKET_DOMAIN_UNIX};
char *protocols[] = {"tcp://", "ipc://"};
const unsigned int protocols_size = GS_SOCKET_DOMAIN_AMOUNT;

unsigned int index;

for (index = 0; index < protocols_size; ++index) {
    if (strncmp(address, protocols[index], strlen(protocols[index])) == 0) {
        break;
    }
}

if (index == protocols_size) {
    print_usage(argv[0]);
    exit(EXIT_FAILURE);
}

send_message(types[index], address + strlen(protocols[index]), message, strlen(message));
```

**Comment by yjfan** [comment]
Hi @joeyoungisme,
Please help to update the code accordingly.

**Comment by yjfan** [comment]
Hi @joeyoungisme,
I have reverted the merge.
Please help to check the comments above and update the code accordingly.
Thank you for your help.

**Comment by joeyoungisme** [comment]
Fix bad practice above.

Changing protocol comparison method better.

**Pull request merged**